{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f2323a6c-d5d0-418b-aa1e-6bda60f416af",
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.request import urlopen\n",
    "\n",
    "from io import TextIOWrapper\n",
    "\n",
    "from Bio import SeqIO \n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import reframed\n",
    "\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "516c5d99-1893-455b-a6d1-a6c79466f286",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Acetate</th>\n",
       "      <th>Propionate</th>\n",
       "      <th>Butyrate</th>\n",
       "      <th>Valerate</th>\n",
       "      <th>Hexanoate</th>\n",
       "      <th>Iso-butyrate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.A.1.120.5</td>\n",
       "      <td>1.A.46.2.1</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "      <td>2.A.73.1.1</td>\n",
       "      <td>1.B.25.1.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.A.14.2.2</td>\n",
       "      <td>2.A.21.4.1</td>\n",
       "      <td>2.A.1.19.18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.A.1.13.1</td>\n",
       "      <td>2.A.21.5.3</td>\n",
       "      <td>2.A.21.4.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.A.1.6.11</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "      <td>2.A.21.5.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.A.102.4.13</td>\n",
       "      <td>2.A.8.1.13</td>\n",
       "      <td>2.A.21.5.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2.A.21.7.3</td>\n",
       "      <td>3.A.1.205.3</td>\n",
       "      <td>2.A.21.5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2.A.23.1.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.A.73.1.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2.A.6.2.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.A.8.1.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2.A.96.1.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2.A.96.1.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Acetate   Propionate     Butyrate    Valerate   Hexanoate  \\\n",
       "0   3.A.1.120.5   1.A.46.2.1   2.A.1.13.1  2.A.73.1.1  1.B.25.1.8   \n",
       "1    1.A.14.2.2   2.A.21.4.1  2.A.1.19.18         NaN         NaN   \n",
       "2    2.A.1.13.1   2.A.21.5.3   2.A.21.4.1         NaN         NaN   \n",
       "3    2.A.1.6.11   2.A.21.7.3   2.A.21.5.3         NaN         NaN   \n",
       "4  2.A.102.4.13   2.A.8.1.13   2.A.21.5.4         NaN         NaN   \n",
       "5    2.A.21.7.3  3.A.1.205.3   2.A.21.5.5         NaN         NaN   \n",
       "6   2.A.23.1.10          NaN   2.A.73.1.1         NaN         NaN   \n",
       "7     2.A.6.2.9          NaN   2.A.8.1.13         NaN         NaN   \n",
       "8   2.A.96.1.11          NaN          NaN         NaN         NaN   \n",
       "9    2.A.96.1.4          NaN          NaN         NaN         NaN   \n",
       "\n",
       "   Iso-butyrate  \n",
       "0           NaN  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  \n",
       "5           NaN  \n",
       "6           NaN  \n",
       "7           NaN  \n",
       "8           NaN  \n",
       "9           NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possible_transporters = pd.read_excel(\"/Users/idunmariaburgos/Documents/Work/Project/chain elongation/Clostridia BL3 and BL4/possible_transporters_tcdb.xlsx\",sheet_name=\"overview\")\n",
    "possible_transporters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30ea037c-05c9-41ef-8bbd-7ef8d223a610",
   "metadata": {},
   "source": [
    "- 3.A.1.120.5 - abc\n",
    "- 1.A.14.2.2 - acetate uptake\n",
    "- 2.A.23.1.10 - uptake\n",
    "- 2.A.6.2.9 - abc\n",
    "- 2.A.96.1.11 - uptake\n",
    "\n",
    "\n",
    "- 2.A.1.13.1 - proton linked efflux/influx\n",
    "- 2.A.1.6.11 - Acetate/haloacid transporter (no mechanism)\n",
    "- 2.A.102.4.13 - acetate transporter\n",
    "- 2.A.21.7.3 - Pyruvate/acetate/propionate: H+ symporter\n",
    "- 2.A.96.1.4 - acetate permease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7d4a818e-2c98-4fa8-9ddc-884775963a6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "possible_transporters = ['2.A.1.13.1','2.A.1.6.11','2.A.102.4.13','2.A.21.7.3','2.A.96.1.4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4cb4b015-83df-4732-af35-da3e9d331edc",
   "metadata": {},
   "outputs": [],
   "source": [
    "possible_transporters_regex = \"|\".join(possible_transporters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "556dd959-f655-47bd-b0fe-3d8526aabe4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2.A.1.13.1|2.A.1.6.11|2.A.102.4.13|2.A.21.7.3|2.A.96.1.4'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possible_transporters_regex"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3153ca83-2813-49d9-abd7-9b6b53a25338",
   "metadata": {},
   "source": [
    "**Filter data from tcdb**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0858227d-a99f-46ed-ac1e-6818926ff6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = TextIOWrapper(urlopen(\"https://www.tcdb.org/public/tcdb\"))\n",
    "recs = list(SeqIO.parse(handle, 'fasta'))\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b6d8461c-f1c1-4142-a180-35926de9b5a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ofile = open(\"transporters/transporters.faa\",\"w\")\n",
    "\n",
    "for rec in recs:\n",
    "    id_ = str(rec.id)\n",
    "    \n",
    "    if re.search(possible_transporters_regex, id_):\n",
    "        \n",
    "        \n",
    "        sequence = str(rec.seq)\n",
    "        \n",
    "        ofile.write(\">\"+id_+\"\\n\"+sequence+\"\\n\")\n",
    "ofile.close()        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d935c2da-3b37-4e2d-8828-16e12b73b353",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "178dd752-7d5c-4734-b66c-1bb60a8a80a4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ACt2r_MAGs = []\n",
    "\n",
    "ACt2r_MAGs_data = []\n",
    "for filename in os.listdir(\"transporters/\"):\n",
    "    \n",
    "    if filename.endswith(\".tsv\"):\n",
    "\n",
    "        transport= pd.read_csv(\"transporters/\"+filename,sep=\"\\t\",header=None)\n",
    "        transport.columns = [\"query acc.ver\", \"subject acc.ver\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\", \"s. start\", \"s. end\", \"evalue\", \"bit score\"]\n",
    "        transport.reset_index(inplace=True)\n",
    "        \n",
    "        if transport.shape[0]>1:\n",
    "            \n",
    "            transport.sort_values(by=\"evalue\",ascending=True,inplace=True)\n",
    "            transport.reset_index(inplace=True)\n",
    "            \n",
    "            query = transport.loc[0,\"query acc.ver\"]\n",
    "            gene = transport.loc[0,\"subject acc.ver\"]\n",
    "            best_evalue = transport.loc[0,\"evalue\"]\n",
    "            bit_score = transport.loc[0,\"bit score\"]\n",
    "            \n",
    "            if best_evalue<1e-5 and bit_score>20:\n",
    "                if \"2.A.1.13.1\" in query:\n",
    "                    ACt2r_MAGs.append(filename[:-4])\n",
    "                    \n",
    "                    ACt2r_MAGs_data.append((filename[:-4],gene,best_evalue,bit_score,\"2.A.1.13.1\"))\n",
    "\n",
    "                elif \"2.A.21.7.3\" in query: \n",
    "                    ACt2r_MAGs.append(filename[:-4])\n",
    "                    ACt2r_MAGs_data.append((filename[:-4],gene,best_evalue,bit_score,\"2.A.21.7.3\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6e0283a1-beae-408b-a22b-a0de2ecb9596",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ACt2r_MAGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6b6beb36-0db1-4d22-b306-6825a61aac7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MAG</th>\n",
       "      <th>gene</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>TCDB_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>CH15-bin.17</td>\n",
       "      <td>gene545</td>\n",
       "      <td>4.900000e-157</td>\n",
       "      <td>456.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CH13-bin.1</td>\n",
       "      <td>gene374</td>\n",
       "      <td>4.910000e-157</td>\n",
       "      <td>456.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>CH7-bin.16</td>\n",
       "      <td>gene2185</td>\n",
       "      <td>8.200000e-156</td>\n",
       "      <td>452.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CH7-bin.4</td>\n",
       "      <td>gene3341</td>\n",
       "      <td>4.480000e-153</td>\n",
       "      <td>445.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CH8-bin.25</td>\n",
       "      <td>gene3105</td>\n",
       "      <td>2.220000e-151</td>\n",
       "      <td>441.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CH1-bin.9</td>\n",
       "      <td>gene769</td>\n",
       "      <td>1.280000e-150</td>\n",
       "      <td>439.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CH15-bin.6</td>\n",
       "      <td>gene1704</td>\n",
       "      <td>1.430000e-142</td>\n",
       "      <td>420.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CH1-bin.1</td>\n",
       "      <td>gene727</td>\n",
       "      <td>3.590000e-56</td>\n",
       "      <td>191.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CH14-bin.2</td>\n",
       "      <td>gene1003</td>\n",
       "      <td>9.490000e-23</td>\n",
       "      <td>99.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CH9-bin.1</td>\n",
       "      <td>gene3888</td>\n",
       "      <td>1.190000e-22</td>\n",
       "      <td>99.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CH13-bin.10</td>\n",
       "      <td>gene2510</td>\n",
       "      <td>4.490000e-19</td>\n",
       "      <td>87.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CH9-bin.8</td>\n",
       "      <td>gene1</td>\n",
       "      <td>1.140000e-18</td>\n",
       "      <td>84.3</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CH15-bin.18</td>\n",
       "      <td>gene948</td>\n",
       "      <td>2.010000e-17</td>\n",
       "      <td>82.4</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CH13-bin.14</td>\n",
       "      <td>gene5876</td>\n",
       "      <td>2.620000e-17</td>\n",
       "      <td>82.4</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CH7-bin.11</td>\n",
       "      <td>gene2382</td>\n",
       "      <td>4.360000e-17</td>\n",
       "      <td>80.5</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>CH8-bin.2</td>\n",
       "      <td>gene979</td>\n",
       "      <td>4.580000e-17</td>\n",
       "      <td>80.5</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CH8-bin.12</td>\n",
       "      <td>gene529</td>\n",
       "      <td>5.030000e-17</td>\n",
       "      <td>80.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CH8-bin.11</td>\n",
       "      <td>gene2414</td>\n",
       "      <td>8.090000e-17</td>\n",
       "      <td>79.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>CH8-bin.4</td>\n",
       "      <td>gene68</td>\n",
       "      <td>8.480000e-17</td>\n",
       "      <td>78.6</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CH8-bin.14</td>\n",
       "      <td>gene546</td>\n",
       "      <td>9.540000e-17</td>\n",
       "      <td>79.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CH13-bin.0</td>\n",
       "      <td>gene2567</td>\n",
       "      <td>1.030000e-16</td>\n",
       "      <td>79.3</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CH7-bin.24</td>\n",
       "      <td>gene385</td>\n",
       "      <td>6.930000e-16</td>\n",
       "      <td>75.5</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CH7-bin.13</td>\n",
       "      <td>gene2195</td>\n",
       "      <td>1.170000e-15</td>\n",
       "      <td>75.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CH13-bin.4</td>\n",
       "      <td>gene2613</td>\n",
       "      <td>1.580000e-15</td>\n",
       "      <td>75.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CH15-bin.1</td>\n",
       "      <td>gene2599</td>\n",
       "      <td>1.620000e-15</td>\n",
       "      <td>75.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CH8-bin.16</td>\n",
       "      <td>gene1114</td>\n",
       "      <td>2.510000e-15</td>\n",
       "      <td>75.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CH13-bin.13</td>\n",
       "      <td>gene458</td>\n",
       "      <td>2.780000e-15</td>\n",
       "      <td>74.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CH8-bin.9</td>\n",
       "      <td>gene614</td>\n",
       "      <td>3.530000e-15</td>\n",
       "      <td>73.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CH3-bin.1</td>\n",
       "      <td>gene4869</td>\n",
       "      <td>3.850000e-15</td>\n",
       "      <td>75.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CH15-bin.23</td>\n",
       "      <td>gene1853</td>\n",
       "      <td>7.300000e-15</td>\n",
       "      <td>73.6</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CH8-bin.27</td>\n",
       "      <td>gene588</td>\n",
       "      <td>7.450000e-15</td>\n",
       "      <td>73.6</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CH7-bin.20</td>\n",
       "      <td>gene898</td>\n",
       "      <td>2.310000e-14</td>\n",
       "      <td>70.9</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CH3-bin.2</td>\n",
       "      <td>gene2757</td>\n",
       "      <td>2.650000e-14</td>\n",
       "      <td>72.4</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CH14-bin.1</td>\n",
       "      <td>gene1344</td>\n",
       "      <td>6.140000e-14</td>\n",
       "      <td>71.2</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CH13-bin.2</td>\n",
       "      <td>gene1742</td>\n",
       "      <td>6.680000e-14</td>\n",
       "      <td>70.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>CH7-bin.8</td>\n",
       "      <td>gene2703</td>\n",
       "      <td>1.350000e-13</td>\n",
       "      <td>69.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>CH7-bin.15</td>\n",
       "      <td>gene1972</td>\n",
       "      <td>3.030000e-13</td>\n",
       "      <td>68.6</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CH1-bin.0</td>\n",
       "      <td>gene2609</td>\n",
       "      <td>3.480000e-13</td>\n",
       "      <td>68.2</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CH15-bin.7</td>\n",
       "      <td>gene621</td>\n",
       "      <td>3.000000e-12</td>\n",
       "      <td>65.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CH13-bin.25</td>\n",
       "      <td>gene855</td>\n",
       "      <td>3.060000e-12</td>\n",
       "      <td>65.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CH1-bin.4</td>\n",
       "      <td>gene2269</td>\n",
       "      <td>8.890000e-12</td>\n",
       "      <td>63.2</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CH13-bin.20</td>\n",
       "      <td>gene1181</td>\n",
       "      <td>4.790000e-11</td>\n",
       "      <td>61.2</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>CH15-bin.10</td>\n",
       "      <td>gene75</td>\n",
       "      <td>7.650000e-11</td>\n",
       "      <td>60.8</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CH7-bin.21</td>\n",
       "      <td>gene1580</td>\n",
       "      <td>1.220000e-10</td>\n",
       "      <td>60.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CH8-bin.0</td>\n",
       "      <td>gene1211</td>\n",
       "      <td>1.460000e-10</td>\n",
       "      <td>58.5</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CH13-bin.16</td>\n",
       "      <td>gene888</td>\n",
       "      <td>1.740000e-10</td>\n",
       "      <td>58.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>CH7-bin.9</td>\n",
       "      <td>gene1470</td>\n",
       "      <td>4.890000e-10</td>\n",
       "      <td>58.5</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>CH1-bin.8</td>\n",
       "      <td>gene2827</td>\n",
       "      <td>1.340000e-08</td>\n",
       "      <td>54.3</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>CH13-bin.24</td>\n",
       "      <td>gene534</td>\n",
       "      <td>2.740000e-08</td>\n",
       "      <td>51.6</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CH7-bin.25</td>\n",
       "      <td>gene808</td>\n",
       "      <td>3.460000e-07</td>\n",
       "      <td>48.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CH8-bin.17</td>\n",
       "      <td>gene51</td>\n",
       "      <td>3.780000e-07</td>\n",
       "      <td>48.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CH15-bin.0</td>\n",
       "      <td>gene1303</td>\n",
       "      <td>6.750000e-07</td>\n",
       "      <td>48.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CH13-bin.17</td>\n",
       "      <td>gene572</td>\n",
       "      <td>8.200000e-07</td>\n",
       "      <td>47.8</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CH8-bin.5</td>\n",
       "      <td>gene1101</td>\n",
       "      <td>8.840000e-07</td>\n",
       "      <td>47.8</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CH15-bin.13</td>\n",
       "      <td>gene1023</td>\n",
       "      <td>4.300000e-06</td>\n",
       "      <td>45.4</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            MAG      gene         evalue  bit_score     TCDB_id\n",
       "43  CH15-bin.17   gene545  4.900000e-157      456.0  2.A.21.7.3\n",
       "12   CH13-bin.1   gene374  4.910000e-157      456.0  2.A.21.7.3\n",
       "40   CH7-bin.16  gene2185  8.200000e-156      452.0  2.A.21.7.3\n",
       "2     CH7-bin.4  gene3341  4.480000e-153      445.0  2.A.21.7.3\n",
       "28   CH8-bin.25  gene3105  2.220000e-151      441.0  2.A.21.7.3\n",
       "41    CH1-bin.9   gene769  1.280000e-150      439.0  2.A.21.7.3\n",
       "13   CH15-bin.6  gene1704  1.430000e-142      420.0  2.A.21.7.3\n",
       "11    CH1-bin.1   gene727   3.590000e-56      191.0  2.A.21.7.3\n",
       "25   CH14-bin.2  gene1003   9.490000e-23       99.0  2.A.21.7.3\n",
       "49    CH9-bin.1  gene3888   1.190000e-22       99.0  2.A.21.7.3\n",
       "18  CH13-bin.10  gene2510   4.490000e-19       87.0  2.A.21.7.3\n",
       "15    CH9-bin.8     gene1   1.140000e-18       84.3  2.A.1.13.1\n",
       "22  CH15-bin.18   gene948   2.010000e-17       82.4  2.A.21.7.3\n",
       "7   CH13-bin.14  gene5876   2.620000e-17       82.4  2.A.21.7.3\n",
       "48   CH7-bin.11  gene2382   4.360000e-17       80.5  2.A.21.7.3\n",
       "39    CH8-bin.2   gene979   4.580000e-17       80.5  2.A.21.7.3\n",
       "37   CH8-bin.12   gene529   5.030000e-17       80.1  2.A.21.7.3\n",
       "38   CH8-bin.11  gene2414   8.090000e-17       79.7  2.A.21.7.3\n",
       "35    CH8-bin.4    gene68   8.480000e-17       78.6  2.A.1.13.1\n",
       "30   CH8-bin.14   gene546   9.540000e-17       79.7  2.A.21.7.3\n",
       "19   CH13-bin.0  gene2567   1.030000e-16       79.3  2.A.21.7.3\n",
       "16   CH7-bin.24   gene385   6.930000e-16       75.5  2.A.1.13.1\n",
       "53   CH7-bin.13  gene2195   1.170000e-15       75.9  2.A.21.7.3\n",
       "6    CH13-bin.4  gene2613   1.580000e-15       75.9  2.A.21.7.3\n",
       "3    CH15-bin.1  gene2599   1.620000e-15       75.9  2.A.21.7.3\n",
       "34   CH8-bin.16  gene1114   2.510000e-15       75.1  2.A.21.7.3\n",
       "23  CH13-bin.13   gene458   2.780000e-15       74.7  2.A.21.7.3\n",
       "26    CH8-bin.9   gene614   3.530000e-15       73.9  2.A.21.7.3\n",
       "32    CH3-bin.1  gene4869   3.850000e-15       75.1  2.A.21.7.3\n",
       "8   CH15-bin.23  gene1853   7.300000e-15       73.6  2.A.21.7.3\n",
       "27   CH8-bin.27   gene588   7.450000e-15       73.6  2.A.21.7.3\n",
       "9    CH7-bin.20   gene898   2.310000e-14       70.9  2.A.1.13.1\n",
       "29    CH3-bin.2  gene2757   2.650000e-14       72.4  2.A.21.7.3\n",
       "24   CH14-bin.1  gene1344   6.140000e-14       71.2  2.A.21.7.3\n",
       "21   CH13-bin.2  gene1742   6.680000e-14       70.9  2.A.21.7.3\n",
       "52    CH7-bin.8  gene2703   1.350000e-13       69.7  2.A.21.7.3\n",
       "42   CH7-bin.15  gene1972   3.030000e-13       68.6  2.A.21.7.3\n",
       "5     CH1-bin.0  gene2609   3.480000e-13       68.2  2.A.21.7.3\n",
       "17   CH15-bin.7   gene621   3.000000e-12       65.1  2.A.21.7.3\n",
       "50  CH13-bin.25   gene855   3.060000e-12       65.1  2.A.21.7.3\n",
       "20    CH1-bin.4  gene2269   8.890000e-12       63.2  2.A.1.13.1\n",
       "44  CH13-bin.20  gene1181   4.790000e-11       61.2  2.A.21.7.3\n",
       "51  CH15-bin.10    gene75   7.650000e-11       60.8  2.A.21.7.3\n",
       "10   CH7-bin.21  gene1580   1.220000e-10       60.1  2.A.21.7.3\n",
       "36    CH8-bin.0  gene1211   1.460000e-10       58.5  2.A.1.13.1\n",
       "4   CH13-bin.16   gene888   1.740000e-10       58.9  2.A.21.7.3\n",
       "54    CH7-bin.9  gene1470   4.890000e-10       58.5  2.A.21.7.3\n",
       "45    CH1-bin.8  gene2827   1.340000e-08       54.3  2.A.21.7.3\n",
       "46  CH13-bin.24   gene534   2.740000e-08       51.6  2.A.1.13.1\n",
       "14   CH7-bin.25   gene808   3.460000e-07       48.9  2.A.21.7.3\n",
       "33   CH8-bin.17    gene51   3.780000e-07       48.9  2.A.21.7.3\n",
       "1    CH15-bin.0  gene1303   6.750000e-07       48.1  2.A.21.7.3\n",
       "0   CH13-bin.17   gene572   8.200000e-07       47.8  2.A.21.7.3\n",
       "31    CH8-bin.5  gene1101   8.840000e-07       47.8  2.A.21.7.3\n",
       "47  CH15-bin.13  gene1023   4.300000e-06       45.4  2.A.1.13.1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(ACt2r_MAGs_data,columns=[\"MAG\",\"gene\",\"evalue\",\"bit_score\",\"TCDB_id\"]).sort_values(\"evalue\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ccca61-cb75-42cb-b445-f9441d0857ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MAGs_RecAndAn2",
   "language": "python",
   "name": "mags_recandan2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
