{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56152bfb-c03e-44e2-8f75-a9faa471076d",
   "metadata": {},
   "source": [
    "# Change from genebank to fasta-format"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d3ef945-fd2a-4570-b89f-ff145fad15c4",
   "metadata": {},
   "source": [
    "Strategy:\n",
    "- Make GEM for all genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "33fa5b34-0a01-4821-9c5d-7ca8157dbc86",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "767f35dc-ca15-4e79-91f1-32e2532580f4",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Testing parsing genbank files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "16892ffd-4e18-41be-b20a-a6c766c7acd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAG_data = [] # (gene_id, database_ref, translation)\n",
    "\n",
    "# For each scaffold in the bin\n",
    "\n",
    "scaffolds = 0\n",
    "for seq_record in SeqIO.parse(\"input/genbank/CH1-bin.0.gbk\",\"genbank\"):\n",
    "    scaffolds = scaffolds +1 \n",
    "    # For each gene sequence in the scaffold\n",
    "    for feature in seq_record.features:\n",
    "        \n",
    "        # If the translation field is in the data\n",
    "        if \"translation\" in feature.qualifiers.keys():\n",
    "            \n",
    "            db_ref = None\n",
    "            \n",
    "            if \"db_xref\" in feature.qualifiers.keys():\n",
    "                db_ref = feature.qualifiers[\"db_xref\"][0]\n",
    "\n",
    "            gene_id = feature.qualifiers[\"gene\"][0]\n",
    "            translation = feature.qualifiers[\"translation\"][0]\n",
    "\n",
    "            MAG_data.append((gene_id,db_ref,translation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2cd6632c-a04c-4643-a4ff-b81ae5895e68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "321"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scaffolds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d09b9015-fa01-40f1-849c-3a9e9e060d11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database_ref</th>\n",
       "      <th>translation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CH1-bin.0_NODE_133_length_76246_cov_6.126393_1</td>\n",
       "      <td>None</td>\n",
       "      <td>YVYSQMIAGRSAPRYGEAKNSWLTGTAAWAFVNISQAILGIQPDFD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CH1-bin.0_NODE_133_length_76246_cov_6.126393_2</td>\n",
       "      <td>ko:K05350</td>\n",
       "      <td>MEFPKQFVWGAASSSYQTEGNRSGRGDSIWDEFCTRPGAIRNNETG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CH1-bin.0_NODE_133_length_76246_cov_6.126393_3</td>\n",
       "      <td>None</td>\n",
       "      <td>MDISTHKRRKLLPAAAAVLFSLSVFSATAFATSNIPPQGQEINAGV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CH1-bin.0_NODE_133_length_76246_cov_6.126393_4</td>\n",
       "      <td>ko:K02030</td>\n",
       "      <td>MKKALSALLLVCAIGFSMGGCSGVSSSSSSPTDPSWNTVRTKDTLV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CH1-bin.0_NODE_133_length_76246_cov_6.126393_5</td>\n",
       "      <td>None</td>\n",
       "      <td>MESVWGTVSNSVNEDYFKGAPMKMSTLLATANQKLSAFIRRRSFSE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3228</th>\n",
       "      <td>CH1-bin.0_k141_4018_length_2502_cov_7.0000_1</td>\n",
       "      <td>None</td>\n",
       "      <td>MKEAKMKSLFCKLKNKLGAAAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3229</th>\n",
       "      <td>CH1-bin.0_k141_4018_length_2502_cov_7.0000_2</td>\n",
       "      <td>None</td>\n",
       "      <td>MPGITLKNGLISYYGNPAGYTEKEKAVVDSIFQNDELTAWLKSRSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3230</th>\n",
       "      <td>CH1-bin.0_k141_4018_length_2502_cov_7.0000_3</td>\n",
       "      <td>None</td>\n",
       "      <td>MKKNQMPQIGLPADACERSGFTDKDTLELHAGQNALVFMKDKMTAL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3231</th>\n",
       "      <td>CH1-bin.0_k141_4018_length_2502_cov_7.0000_4</td>\n",
       "      <td>ko:K06223</td>\n",
       "      <td>MSWIGGKKSLRELIVTLFPLYYERYIEVFGGGGWVLFHKLPGNDFE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3232</th>\n",
       "      <td>CH1-bin.0_k141_4018_length_2502_cov_7.0000_5</td>\n",
       "      <td>None</td>\n",
       "      <td>NGRKGGPMQKEQLAEHLRRYHRGAENAVISRELECTFHVSGKELRD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3233 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             gene_id database_ref  \\\n",
       "0     CH1-bin.0_NODE_133_length_76246_cov_6.126393_1         None   \n",
       "1     CH1-bin.0_NODE_133_length_76246_cov_6.126393_2    ko:K05350   \n",
       "2     CH1-bin.0_NODE_133_length_76246_cov_6.126393_3         None   \n",
       "3     CH1-bin.0_NODE_133_length_76246_cov_6.126393_4    ko:K02030   \n",
       "4     CH1-bin.0_NODE_133_length_76246_cov_6.126393_5         None   \n",
       "...                                              ...          ...   \n",
       "3228    CH1-bin.0_k141_4018_length_2502_cov_7.0000_1         None   \n",
       "3229    CH1-bin.0_k141_4018_length_2502_cov_7.0000_2         None   \n",
       "3230    CH1-bin.0_k141_4018_length_2502_cov_7.0000_3         None   \n",
       "3231    CH1-bin.0_k141_4018_length_2502_cov_7.0000_4    ko:K06223   \n",
       "3232    CH1-bin.0_k141_4018_length_2502_cov_7.0000_5         None   \n",
       "\n",
       "                                            translation  \n",
       "0     YVYSQMIAGRSAPRYGEAKNSWLTGTAAWAFVNISQAILGIQPDFD...  \n",
       "1     MEFPKQFVWGAASSSYQTEGNRSGRGDSIWDEFCTRPGAIRNNETG...  \n",
       "2     MDISTHKRRKLLPAAAAVLFSLSVFSATAFATSNIPPQGQEINAGV...  \n",
       "3     MKKALSALLLVCAIGFSMGGCSGVSSSSSSPTDPSWNTVRTKDTLV...  \n",
       "4     MESVWGTVSNSVNEDYFKGAPMKMSTLLATANQKLSAFIRRRSFSE...  \n",
       "...                                                 ...  \n",
       "3228                             MKEAKMKSLFCKLKNKLGAAAV  \n",
       "3229  MPGITLKNGLISYYGNPAGYTEKEKAVVDSIFQNDELTAWLKSRSL...  \n",
       "3230  MKKNQMPQIGLPADACERSGFTDKDTLELHAGQNALVFMKDKMTAL...  \n",
       "3231  MSWIGGKKSLRELIVTLFPLYYERYIEVFGGGGWVLFHKLPGNDFE...  \n",
       "3232  NGRKGGPMQKEQLAEHLRRYHRGAENAVISRELECTFHVSGKELRD...  \n",
       "\n",
       "[3233 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MAG_data_df = pd.DataFrame(MAG_data,columns=[\"gene_id\", \"database_ref\", \"translation\"])\n",
    "MAG_data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "f8afff1a-509f-4b51-bdb7-27507aa52070",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3233"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(MAG_data_df.translation.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "e6cdfaab-51bf-49b2-b361-ee6606507da5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3233"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(MAG_data_df.gene_id.values))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4cf6b83-44b3-4bc0-ac8e-e517325f3b50",
   "metadata": {},
   "source": [
    "### Make fasta file from genbank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "7cbb37ff-1784-4c82-989c-d333f3c13db8",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CH15-bin.17.gbk\n",
      "CH9-bin.4.gbk\n",
      "CH7-bin.14.gbk\n",
      "CH1-bin.8.gbk\n",
      "CH13-bin.20.gbk\n",
      "CH13-bin.21.gbk\n",
      "CH1-bin.9.gbk\n",
      "CH7-bin.15.gbk\n",
      "CH9-bin.5.gbk\n",
      "CH15-bin.16.gbk\n",
      "CH15-bin.8.gbk\n",
      "CH7-bin.17.gbk\n",
      "CH13-bin.22.gbk\n",
      "CH7-bin.16.gbk\n",
      "CH9-bin.6.gbk\n",
      "CH15-bin.15.gbk\n",
      "CH7-bin.9.gbk\n",
      "CH15-bin.11.gbk\n",
      "CH7-bin.12.gbk\n",
      "CH9-bin.2.gbk\n",
      "CH7-bin.13.gbk\n",
      "CH15-bin.10.gbk\n",
      "CH7-bin.8.gbk\n",
      "CH15-bin.12.gbk\n",
      "CH7-bin.11.gbk\n",
      "CH9-bin.1.gbk\n",
      "CH13-bin.19.gbk\n",
      "CH13-bin.25.gbk\n",
      "CH13-bin.24.gbk\n",
      "CH13-bin.8.gbk\n",
      "CH9-bin.0.gbk\n",
      "CH15-bin.13.gbk\n",
      "CH8-bin.16.gbk\n",
      "CH3-bin.0.gbk\n",
      "CH8-bin.4.gbk\n",
      "CH8-bin.5.gbk\n",
      "CH3-bin.1.gbk\n",
      "CH8-bin.17.gbk\n",
      "CH8-bin.29.gbk\n",
      "CH8-bin.7.gbk\n",
      "CH8-bin.6.gbk\n",
      "CH3-bin.2.gbk\n",
      "CH8-bin.14.gbk\n",
      "CH8-bin.2.gbk\n",
      "CH8-bin.11.gbk\n",
      "CH8-bin.0.gbk\n",
      "CH8-bin.12.gbk\n",
      "CH14-bin.2.gbk\n",
      "CH8-bin.22.gbk\n",
      "CH8-bin.20.gbk\n",
      "CH14-bin.1.gbk\n",
      "CH8-bin.21.gbk\n",
      "CH8-bin.25.gbk\n",
      "CH14-bin.4.gbk\n",
      "CH8-bin.8.gbk\n",
      "CH8-bin.9.gbk\n",
      "CH8-bin.27.gbk\n",
      "CH7-bin.6.gbk\n",
      "CH15-bin.22.gbk\n",
      "CH15-bin.2.gbk\n",
      "CH7-bin.21.gbk\n",
      "CH1-bin.1.gbk\n",
      "CH13-bin.14.gbk\n",
      "CH1-bin.0.gbk\n",
      "CH13-bin.4.gbk\n",
      "CH7-bin.20.gbk\n",
      "CH15-bin.3.gbk\n",
      "CH15-bin.23.gbk\n",
      "CH15-bin.1.gbk\n",
      "CH13-bin.6.gbk\n",
      "CH1-bin.2.gbk\n",
      "CH13-bin.16.gbk\n",
      "CH13-bin.17.gbk\n",
      "CH1-bin.3.gbk\n",
      "CH7-bin.23.gbk\n",
      "CH15-bin.0.gbk\n",
      "CH15-bin.20.gbk\n",
      "CH7-bin.4.gbk\n",
      "CH15-bin.4.gbk\n",
      "CH15-bin.18.gbk\n",
      "CH1-bin.7.gbk\n",
      "CH1-bin.10.gbk\n",
      "CH13-bin.13.gbk\n",
      "CH13-bin.12.gbk\n",
      "CH1-bin.6.gbk\n",
      "CH13-bin.2.gbk\n",
      "CH15-bin.5.gbk\n",
      "CH7-bin.1.gbk\n",
      "CH15-bin.7.gbk\n",
      "CH7-bin.24.gbk\n",
      "CH9-bin.8.gbk\n",
      "CH7-bin.18.gbk\n",
      "CH13-bin.0.gbk\n",
      "CH1-bin.4.gbk\n",
      "CH13-bin.10.gbk\n",
      "CH13-bin.11.gbk\n",
      "CH1-bin.5.gbk\n",
      "CH13-bin.1.gbk\n",
      "CH7-bin.25.gbk\n",
      "CH15-bin.6.gbk\n",
      "CH7-bin.2.gbk\n"
     ]
    }
   ],
   "source": [
    "directory_in_str = \"input/genbank/\"\n",
    "directory = os.fsencode(directory_in_str)\n",
    "    \n",
    "gene_ids = {} \n",
    "\n",
    "# For each mag\n",
    "for file in os.listdir(directory):\n",
    "    filename = os.fsdecode(file)\n",
    "\n",
    "    if filename.endswith(\".gbk\"): \n",
    "\n",
    "        print(filename)\n",
    "        \n",
    "        gene_id_new = 0\n",
    "        gene_ids[(filename[:-4],\"old_id\")] = []\n",
    "        gene_ids[(filename[:-4],\"new_id\")] = []\n",
    "        ofile = open(\"output/MAGs_fasta/\"+filename[:-4]+\".faa\",\"w\")\n",
    "        \n",
    "        # For each scaffold\n",
    "        for seq_record in SeqIO.parse(\"input/genbank/\"+filename,\"genbank\"):\n",
    "\n",
    "            # For each gene sequence in the scaffold\n",
    "            for feature in seq_record.features:\n",
    "\n",
    "                # If the translation field is in the data -> could be enzyme\n",
    "                if \"translation\" in feature.qualifiers.keys():\n",
    "                    translation = feature.qualifiers[\"translation\"][0]\n",
    "                    gene_id_old = feature.qualifiers[\"gene\"][0]\n",
    "                    \n",
    "                    \n",
    "                    \n",
    "                    ofile.write(\">\"+\"gene\"+str(gene_id_new)+\"\\n\"+translation+\"\\n\")\n",
    "                    \n",
    "                    gene_ids[(filename[:-4],\"old_id\")].append(gene_id_old)\n",
    "                    gene_ids[(filename[:-4],\"new_id\")].append(\"gene\"+str(gene_id_new))\n",
    "                    \n",
    "                    gene_id_new = gene_id_new+1\n",
    "        ofile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "fe4c1094-13c7-4dc7-ad86-bebb648d2493",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH15-bin.17</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH9-bin.4</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH7-bin.14</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH1-bin.8</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH13-bin.20</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH1-bin.5</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH13-bin.1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH7-bin.25</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH15-bin.6</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CH7-bin.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>...</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "      <th>old_id</th>\n",
       "      <th>new_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_1</td>\n",
       "      <td>gene0</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_1</td>\n",
       "      <td>gene0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_2</td>\n",
       "      <td>gene1</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_2</td>\n",
       "      <td>gene1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_3</td>\n",
       "      <td>gene2</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_3</td>\n",
       "      <td>gene2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_4</td>\n",
       "      <td>gene3</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_4</td>\n",
       "      <td>gene3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_5</td>\n",
       "      <td>gene4</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_5</td>\n",
       "      <td>gene4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_6</td>\n",
       "      <td>gene5</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_6</td>\n",
       "      <td>gene5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_7</td>\n",
       "      <td>gene6</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_7</td>\n",
       "      <td>gene6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_8</td>\n",
       "      <td>gene7</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_8</td>\n",
       "      <td>gene7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.159582_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_9</td>\n",
       "      <td>gene8</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_9</td>\n",
       "      <td>gene8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CH15-bin.17_NODE_1_length_1041084_cov_76.15958...</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH9-bin.4_NODE_5_length_306115_cov_17.353493_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH7-bin.14_NODE_40_length_181369_cov_8.935609_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH1-bin.8_NODE_4_length_584485_cov_43.940039_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH13-bin.20_NODE_283_length_49011_cov_7.985906_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>...</td>\n",
       "      <td>CH1-bin.5_NODE_124_length_82203_cov_6.596983_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH13-bin.1_NODE_1_length_928262_cov_19.438597_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH7-bin.25_NODE_53_length_154058_cov_7.769940_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH15-bin.6_NODE_6_length_825317_cov_75.898953_10</td>\n",
       "      <td>gene9</td>\n",
       "      <td>CH7-bin.2_NODE_4_length_505470_cov_55.993093_10</td>\n",
       "      <td>gene9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         CH15-bin.17         \\\n",
       "                                              old_id new_id   \n",
       "0  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_1  gene0   \n",
       "1  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_2  gene1   \n",
       "2  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_3  gene2   \n",
       "3  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_4  gene3   \n",
       "4  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_5  gene4   \n",
       "5  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_6  gene5   \n",
       "6  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_7  gene6   \n",
       "7  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_8  gene7   \n",
       "8  CH15-bin.17_NODE_1_length_1041084_cov_76.159582_9  gene8   \n",
       "9  CH15-bin.17_NODE_1_length_1041084_cov_76.15958...  gene9   \n",
       "\n",
       "                                         CH9-bin.4         \\\n",
       "                                            old_id new_id   \n",
       "0   CH9-bin.4_NODE_5_length_306115_cov_17.353493_1  gene0   \n",
       "1   CH9-bin.4_NODE_5_length_306115_cov_17.353493_2  gene1   \n",
       "2   CH9-bin.4_NODE_5_length_306115_cov_17.353493_3  gene2   \n",
       "3   CH9-bin.4_NODE_5_length_306115_cov_17.353493_4  gene3   \n",
       "4   CH9-bin.4_NODE_5_length_306115_cov_17.353493_5  gene4   \n",
       "5   CH9-bin.4_NODE_5_length_306115_cov_17.353493_6  gene5   \n",
       "6   CH9-bin.4_NODE_5_length_306115_cov_17.353493_7  gene6   \n",
       "7   CH9-bin.4_NODE_5_length_306115_cov_17.353493_8  gene7   \n",
       "8   CH9-bin.4_NODE_5_length_306115_cov_17.353493_9  gene8   \n",
       "9  CH9-bin.4_NODE_5_length_306115_cov_17.353493_10  gene9   \n",
       "\n",
       "                                         CH7-bin.14         \\\n",
       "                                             old_id new_id   \n",
       "0   CH7-bin.14_NODE_40_length_181369_cov_8.935609_1  gene0   \n",
       "1   CH7-bin.14_NODE_40_length_181369_cov_8.935609_2  gene1   \n",
       "2   CH7-bin.14_NODE_40_length_181369_cov_8.935609_3  gene2   \n",
       "3   CH7-bin.14_NODE_40_length_181369_cov_8.935609_4  gene3   \n",
       "4   CH7-bin.14_NODE_40_length_181369_cov_8.935609_5  gene4   \n",
       "5   CH7-bin.14_NODE_40_length_181369_cov_8.935609_6  gene5   \n",
       "6   CH7-bin.14_NODE_40_length_181369_cov_8.935609_7  gene6   \n",
       "7   CH7-bin.14_NODE_40_length_181369_cov_8.935609_8  gene7   \n",
       "8   CH7-bin.14_NODE_40_length_181369_cov_8.935609_9  gene8   \n",
       "9  CH7-bin.14_NODE_40_length_181369_cov_8.935609_10  gene9   \n",
       "\n",
       "                                         CH1-bin.8         \\\n",
       "                                            old_id new_id   \n",
       "0   CH1-bin.8_NODE_4_length_584485_cov_43.940039_1  gene0   \n",
       "1   CH1-bin.8_NODE_4_length_584485_cov_43.940039_2  gene1   \n",
       "2   CH1-bin.8_NODE_4_length_584485_cov_43.940039_3  gene2   \n",
       "3   CH1-bin.8_NODE_4_length_584485_cov_43.940039_4  gene3   \n",
       "4   CH1-bin.8_NODE_4_length_584485_cov_43.940039_5  gene4   \n",
       "5   CH1-bin.8_NODE_4_length_584485_cov_43.940039_6  gene5   \n",
       "6   CH1-bin.8_NODE_4_length_584485_cov_43.940039_7  gene6   \n",
       "7   CH1-bin.8_NODE_4_length_584485_cov_43.940039_8  gene7   \n",
       "8   CH1-bin.8_NODE_4_length_584485_cov_43.940039_9  gene8   \n",
       "9  CH1-bin.8_NODE_4_length_584485_cov_43.940039_10  gene9   \n",
       "\n",
       "                                         CH13-bin.20         ...  \\\n",
       "                                              old_id new_id  ...   \n",
       "0   CH13-bin.20_NODE_283_length_49011_cov_7.985906_1  gene0  ...   \n",
       "1   CH13-bin.20_NODE_283_length_49011_cov_7.985906_2  gene1  ...   \n",
       "2   CH13-bin.20_NODE_283_length_49011_cov_7.985906_3  gene2  ...   \n",
       "3   CH13-bin.20_NODE_283_length_49011_cov_7.985906_4  gene3  ...   \n",
       "4   CH13-bin.20_NODE_283_length_49011_cov_7.985906_5  gene4  ...   \n",
       "5   CH13-bin.20_NODE_283_length_49011_cov_7.985906_6  gene5  ...   \n",
       "6   CH13-bin.20_NODE_283_length_49011_cov_7.985906_7  gene6  ...   \n",
       "7   CH13-bin.20_NODE_283_length_49011_cov_7.985906_8  gene7  ...   \n",
       "8   CH13-bin.20_NODE_283_length_49011_cov_7.985906_9  gene8  ...   \n",
       "9  CH13-bin.20_NODE_283_length_49011_cov_7.985906_10  gene9  ...   \n",
       "\n",
       "                                         CH1-bin.5         \\\n",
       "                                            old_id new_id   \n",
       "0   CH1-bin.5_NODE_124_length_82203_cov_6.596983_1  gene0   \n",
       "1   CH1-bin.5_NODE_124_length_82203_cov_6.596983_2  gene1   \n",
       "2   CH1-bin.5_NODE_124_length_82203_cov_6.596983_3  gene2   \n",
       "3   CH1-bin.5_NODE_124_length_82203_cov_6.596983_4  gene3   \n",
       "4   CH1-bin.5_NODE_124_length_82203_cov_6.596983_5  gene4   \n",
       "5   CH1-bin.5_NODE_124_length_82203_cov_6.596983_6  gene5   \n",
       "6   CH1-bin.5_NODE_124_length_82203_cov_6.596983_7  gene6   \n",
       "7   CH1-bin.5_NODE_124_length_82203_cov_6.596983_8  gene7   \n",
       "8   CH1-bin.5_NODE_124_length_82203_cov_6.596983_9  gene8   \n",
       "9  CH1-bin.5_NODE_124_length_82203_cov_6.596983_10  gene9   \n",
       "\n",
       "                                         CH13-bin.1         \\\n",
       "                                             old_id new_id   \n",
       "0   CH13-bin.1_NODE_1_length_928262_cov_19.438597_1  gene0   \n",
       "1   CH13-bin.1_NODE_1_length_928262_cov_19.438597_2  gene1   \n",
       "2   CH13-bin.1_NODE_1_length_928262_cov_19.438597_3  gene2   \n",
       "3   CH13-bin.1_NODE_1_length_928262_cov_19.438597_4  gene3   \n",
       "4   CH13-bin.1_NODE_1_length_928262_cov_19.438597_5  gene4   \n",
       "5   CH13-bin.1_NODE_1_length_928262_cov_19.438597_6  gene5   \n",
       "6   CH13-bin.1_NODE_1_length_928262_cov_19.438597_7  gene6   \n",
       "7   CH13-bin.1_NODE_1_length_928262_cov_19.438597_8  gene7   \n",
       "8   CH13-bin.1_NODE_1_length_928262_cov_19.438597_9  gene8   \n",
       "9  CH13-bin.1_NODE_1_length_928262_cov_19.438597_10  gene9   \n",
       "\n",
       "                                         CH7-bin.25         \\\n",
       "                                             old_id new_id   \n",
       "0   CH7-bin.25_NODE_53_length_154058_cov_7.769940_1  gene0   \n",
       "1   CH7-bin.25_NODE_53_length_154058_cov_7.769940_2  gene1   \n",
       "2   CH7-bin.25_NODE_53_length_154058_cov_7.769940_3  gene2   \n",
       "3   CH7-bin.25_NODE_53_length_154058_cov_7.769940_4  gene3   \n",
       "4   CH7-bin.25_NODE_53_length_154058_cov_7.769940_5  gene4   \n",
       "5   CH7-bin.25_NODE_53_length_154058_cov_7.769940_6  gene5   \n",
       "6   CH7-bin.25_NODE_53_length_154058_cov_7.769940_7  gene6   \n",
       "7   CH7-bin.25_NODE_53_length_154058_cov_7.769940_8  gene7   \n",
       "8   CH7-bin.25_NODE_53_length_154058_cov_7.769940_9  gene8   \n",
       "9  CH7-bin.25_NODE_53_length_154058_cov_7.769940_10  gene9   \n",
       "\n",
       "                                         CH15-bin.6         \\\n",
       "                                             old_id new_id   \n",
       "0   CH15-bin.6_NODE_6_length_825317_cov_75.898953_1  gene0   \n",
       "1   CH15-bin.6_NODE_6_length_825317_cov_75.898953_2  gene1   \n",
       "2   CH15-bin.6_NODE_6_length_825317_cov_75.898953_3  gene2   \n",
       "3   CH15-bin.6_NODE_6_length_825317_cov_75.898953_4  gene3   \n",
       "4   CH15-bin.6_NODE_6_length_825317_cov_75.898953_5  gene4   \n",
       "5   CH15-bin.6_NODE_6_length_825317_cov_75.898953_6  gene5   \n",
       "6   CH15-bin.6_NODE_6_length_825317_cov_75.898953_7  gene6   \n",
       "7   CH15-bin.6_NODE_6_length_825317_cov_75.898953_8  gene7   \n",
       "8   CH15-bin.6_NODE_6_length_825317_cov_75.898953_9  gene8   \n",
       "9  CH15-bin.6_NODE_6_length_825317_cov_75.898953_10  gene9   \n",
       "\n",
       "                                         CH7-bin.2         \n",
       "                                            old_id new_id  \n",
       "0   CH7-bin.2_NODE_4_length_505470_cov_55.993093_1  gene0  \n",
       "1   CH7-bin.2_NODE_4_length_505470_cov_55.993093_2  gene1  \n",
       "2   CH7-bin.2_NODE_4_length_505470_cov_55.993093_3  gene2  \n",
       "3   CH7-bin.2_NODE_4_length_505470_cov_55.993093_4  gene3  \n",
       "4   CH7-bin.2_NODE_4_length_505470_cov_55.993093_5  gene4  \n",
       "5   CH7-bin.2_NODE_4_length_505470_cov_55.993093_6  gene5  \n",
       "6   CH7-bin.2_NODE_4_length_505470_cov_55.993093_7  gene6  \n",
       "7   CH7-bin.2_NODE_4_length_505470_cov_55.993093_8  gene7  \n",
       "8   CH7-bin.2_NODE_4_length_505470_cov_55.993093_9  gene8  \n",
       "9  CH7-bin.2_NODE_4_length_505470_cov_55.993093_10  gene9  \n",
       "\n",
       "[10 rows x 202 columns]"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ids_df = pd.DataFrame.from_dict({ key:pd.Series(value) for key, value in gene_ids.items()})\n",
    "\n",
    "gene_ids_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abbef723-a9be-4e7e-b5e2-b4dc9d20c28d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RecAndAn",
   "language": "python",
   "name": "recandan"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
