{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "076065f0-f656-4477-90e0-6ce1fd199ff1",
   "metadata": {},
   "source": [
    "### Changing acetate transporters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1591d0e0-c07d-4956-9f01-b2b2473eb7e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import reframed\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import os\n",
    "import json\n",
    "\n",
    "import copy\n",
    "\n",
    "from reframed.io.sbml import parse_gpr_rule\n",
    "import functions.general_functions as general_func"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f6ac0d8-efc2-400d-b14f-89497bbc33ad",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Load data"
   ]
  },
  {
   "cell_type": "raw",
   "id": "68d3e96f-4e9b-4d22-95ee-85a11483963e",
   "metadata": {},
   "source": [
    "all_mags_paper = pd.read_excel(\"input/files_from_fairdomhub/All_Mags_for_paper_analysis.xlsx\",sheet_name=\"Coverage\")\n",
    "all_mags_paper.set_index(\"MAG\",inplace=True)\n",
    "all_mags_paper.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "270bd628-8891-4fd8-b1d9-464f0c03048b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Substrate</th>\n",
       "      <th>Completeness (%)</th>\n",
       "      <th>Contamination (%)</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genome size (bp)</th>\n",
       "      <th>Scaffolds</th>\n",
       "      <th>N50</th>\n",
       "      <th>CDS prediction (DRAM)</th>\n",
       "      <th>Column1</th>\n",
       "      <th>new_coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CH14-bin.1</th>\n",
       "      <td>Compost_Digestate</td>\n",
       "      <td>Xylan</td>\n",
       "      <td>99.37</td>\n",
       "      <td>0.32</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lacrimispora</td>\n",
       "      <td>Lacrimispora amygdalina_A</td>\n",
       "      <td>5147597.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>197896.0</td>\n",
       "      <td>4902.0</td>\n",
       "      <td>CH14-bin.1</td>\n",
       "      <td>54.707294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH14-bin.2</th>\n",
       "      <td>Compost_Digestate</td>\n",
       "      <td>Xylan</td>\n",
       "      <td>99.37</td>\n",
       "      <td>0.63</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Lachnospirales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Robinsoniella</td>\n",
       "      <td>Robinsoniella peoriensis</td>\n",
       "      <td>5647052.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>172899.0</td>\n",
       "      <td>4618.0</td>\n",
       "      <td>CH14-bin.2</td>\n",
       "      <td>18.831114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Source Substrate  Completeness (%)  Contamination (%)  \\\n",
       "MAG                                                                            \n",
       "CH14-bin.1  Compost_Digestate     Xylan             99.37               0.32   \n",
       "CH14-bin.2  Compost_Digestate     Xylan             99.37               0.63   \n",
       "\n",
       "              Domain      Phylum       Class           Order           Family  \\\n",
       "MAG                                                                             \n",
       "CH14-bin.1  Bacteria  Firmicutes  Clostridia  Lachnospirales  Lachnospiraceae   \n",
       "CH14-bin.2  Bacteria  Firmicutes  Clostridia  Lachnospirales  Lachnospiraceae   \n",
       "\n",
       "                    Genus                    Species  Genome size (bp)  \\\n",
       "MAG                                                                      \n",
       "CH14-bin.1   Lacrimispora  Lacrimispora amygdalina_A         5147597.0   \n",
       "CH14-bin.2  Robinsoniella   Robinsoniella peoriensis         5647052.0   \n",
       "\n",
       "            Scaffolds       N50  CDS prediction (DRAM)     Column1  \\\n",
       "MAG                                                                  \n",
       "CH14-bin.1       48.0  197896.0                 4902.0  CH14-bin.1   \n",
       "CH14-bin.2       70.0  172899.0                 4618.0  CH14-bin.2   \n",
       "\n",
       "            new_coverage  \n",
       "MAG                       \n",
       "CH14-bin.1     54.707294  \n",
       "CH14-bin.2     18.831114  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_mags_paper = general_func.read_allmags_data()\n",
    "all_mags_paper.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "83d75166-a1a7-44f6-a412-529003055ca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"output/relevant_MAGs_99.txt\") as text_file:\n",
    "    relevant_MAGs = text_file.read().split(\"\\n\")\n",
    "\n",
    "relevant_MAGs = [string.replace(\"\\t\",\"\") for string in relevant_MAGs]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba1a996c-3e5c-4933-833e-688c6524c6b8",
   "metadata": {},
   "source": [
    "#### Load models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "23ee2ed9-0772-4bc0-9c73-ffe4e5ef313d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CH7-bin.18.xml\n",
      "CH15-bin.7.xml\n",
      "CH13-bin.0.xml\n",
      "CH1-bin.4.xml\n",
      "CH13-bin.1.xml\n",
      "CH13-bin.11.xml\n",
      "CH15-bin.6.xml\n",
      "CH7-bin.2.xml\n",
      "CH13-bin.13.xml\n",
      "CH1-bin.10.xml\n",
      "CH1-bin.6.xml\n",
      "CH13-bin.2.xml\n",
      "CH13-bin.12.xml\n",
      "CH15-bin.5.xml\n",
      "CH7-bin.1.xml\n",
      "CH15-bin.1.xml\n",
      "CH1-bin.3.xml\n",
      "CH13-bin.17.xml\n",
      "CH15-bin.0.xml\n",
      "CH7-bin.4.xml\n",
      "CH7-bin.23.xml\n",
      "CH7-bin.6.xml\n",
      "CH15-bin.22.xml\n",
      "CH15-bin.2.xml\n",
      "CH1-bin.1.xml\n",
      "CH13-bin.4.xml\n",
      "CH13-bin.14.xml\n",
      "CH15-bin.23.xml\n",
      "CH7-bin.20.xml\n",
      "CH8-bin.8.xml\n",
      "CH8-bin.9.xml\n",
      "CH8-bin.25.xml\n",
      "CH14-bin.4.xml\n",
      "CH14-bin.1.xml\n",
      "CH8-bin.21.xml\n",
      "CH14-bin.2.xml\n",
      "CH8-bin.22.xml\n",
      "CH8-bin.2.xml\n",
      "CH8-bin.29.xml\n",
      "CH8-bin.7.xml\n",
      "CH8-bin.6.xml\n",
      "CH3-bin.2.xml\n",
      "CH8-bin.14.xml\n",
      "CH8-bin.16.xml\n",
      "CH3-bin.0.xml\n",
      "CH8-bin.5.xml\n",
      "CH3-bin.1.xml\n",
      "CH8-bin.17.xml\n",
      "CH7-bin.11.xml\n",
      "CH9-bin.1.xml\n",
      "CH15-bin.12.xml\n",
      "CH13-bin.25.xml\n",
      "CH15-bin.13.xml\n",
      "CH9-bin.0.xml\n",
      "CH7-bin.12.xml\n",
      "CH9-bin.2.xml\n",
      "CH7-bin.9.xml\n",
      "CH15-bin.10.xml\n",
      "CH7-bin.8.xml\n",
      "CH7-bin.13.xml\n",
      "CH7-bin.17.xml\n",
      "CH15-bin.8.xml\n",
      "CH15-bin.15.xml\n",
      "CH7-bin.16.xml\n",
      "CH9-bin.6.xml\n",
      "CH9-bin.4.xml\n",
      "CH15-bin.17.xml\n",
      "CH1-bin.8.xml\n",
      "CH1-bin.9.xml\n",
      "CH15-bin.16.xml\n",
      "CH7-bin.15.xml\n",
      "CH9-bin.5.xml\n"
     ]
    }
   ],
   "source": [
    "GEMs_dict = {}\n",
    "\n",
    "directory = os.fsencode(\"output/GEMs/GEMs_adapt_media/\")\n",
    "\n",
    "for file in os.listdir(directory):\n",
    "    filename = os.fsdecode(file)\n",
    "    print(filename)\n",
    "\n",
    "    GEMs_dict[filename[:-4]]= reframed.load_cbmodel(\"output/GEMs/GEMs_adapt_media/\"+filename)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bfb0f45-bd91-41e6-9aef-c300565bbd71",
   "metadata": {},
   "source": [
    "## Find acetate producers/consumers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e9c445c-4f0f-452e-907d-6fe6e67409e2",
   "metadata": {},
   "source": [
    "#### Find which models has GPRs for producing acetate through phosphotransacetylase(R_PTAr) and acetate kinase (R_ACKr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1e28af73-520d-45f1-b43a-91980444d1c1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "has_enzymes_for_acetate = {}\n",
    "for MAG in GEMs_dict.keys():\n",
    "    \n",
    "    has_enzyme = []\n",
    "    if \"R_ACKr\" in GEMs_dict[MAG].reactions:\n",
    "        if GEMs_dict[MAG].reactions[\"R_ACKr\"].gpr!=None:\n",
    "            has_enzyme.append(\"R_ACKr\")\n",
    "    if \"R_PTAr\" in GEMs_dict[MAG].reactions:\n",
    "        \n",
    "        if GEMs_dict[MAG].reactions[\"R_PTAr\"].gpr!=None:\n",
    "            has_enzyme.append(\"R_PTAr\")\n",
    "    \n",
    "    has_enzymes_for_acetate[MAG]= len(has_enzyme)==2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b80ff4ad-75bb-4c38-952d-ec21f70fc7f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "has_enzymes_for_acetate_MAGs = list(pd.Series(has_enzymes_for_acetate)[pd.Series(has_enzymes_for_acetate)].index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d324df-bff6-4526-921b-5965d67f8dd9",
   "metadata": {},
   "source": [
    "#### Find hits on the ACt2r protein from the TCDB database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "178dd752-7d5c-4734-b66c-1bb60a8a80a4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ACt2r_MAGs = []\n",
    "\n",
    "ACt2r_MAGs_data = []\n",
    "for filename in os.listdir(\"transporters/\"):\n",
    "    \n",
    "    if filename.endswith(\".tsv\"):\n",
    "\n",
    "        transport= pd.read_csv(\"transporters/\"+filename,sep=\"\\t\",header=None)\n",
    "        transport.columns = [\"query acc.ver\", \"subject acc.ver\", \"% identity\", \"alignment length\", \"mismatches\", \"gap opens\", \"q. start\", \"q. end\", \"s. start\", \"s. end\", \"evalue\", \"bit score\"]\n",
    "        transport.reset_index(inplace=True)\n",
    "        \n",
    "        if transport.shape[0]>1:\n",
    "            \n",
    "            transport.sort_values(by=\"evalue\",ascending=True,inplace=True)\n",
    "            transport.reset_index(inplace=True)\n",
    "            \n",
    "            query = transport.loc[0,\"query acc.ver\"]\n",
    "            gene = transport.loc[0,\"subject acc.ver\"]\n",
    "            best_evalue = transport.loc[0,\"evalue\"]\n",
    "            bit_score = transport.loc[0,\"bit score\"]\n",
    "            \n",
    "            if best_evalue<1e-5 and bit_score>20:\n",
    "                if \"2.A.1.13.1\" in query:\n",
    "                    ACt2r_MAGs.append(filename[:-4])\n",
    "                    \n",
    "                    ACt2r_MAGs_data.append((filename[:-4],gene,best_evalue,bit_score,\"2.A.1.13.1\"))\n",
    "\n",
    "                elif \"2.A.21.7.3\" in query: \n",
    "                    ACt2r_MAGs.append(filename[:-4])\n",
    "                    ACt2r_MAGs_data.append((filename[:-4],gene,best_evalue,bit_score,\"2.A.21.7.3\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6b6beb36-0db1-4d22-b306-6825a61aac7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>TCDB_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAG</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CH15-bin.17</th>\n",
       "      <td>gene545</td>\n",
       "      <td>4.900000e-157</td>\n",
       "      <td>456.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.1</th>\n",
       "      <td>gene374</td>\n",
       "      <td>4.910000e-157</td>\n",
       "      <td>456.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.16</th>\n",
       "      <td>gene2185</td>\n",
       "      <td>8.200000e-156</td>\n",
       "      <td>452.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.4</th>\n",
       "      <td>gene3341</td>\n",
       "      <td>4.480000e-153</td>\n",
       "      <td>445.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.25</th>\n",
       "      <td>gene3105</td>\n",
       "      <td>2.220000e-151</td>\n",
       "      <td>441.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH1-bin.9</th>\n",
       "      <td>gene769</td>\n",
       "      <td>1.280000e-150</td>\n",
       "      <td>439.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.6</th>\n",
       "      <td>gene1704</td>\n",
       "      <td>1.430000e-142</td>\n",
       "      <td>420.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH1-bin.1</th>\n",
       "      <td>gene727</td>\n",
       "      <td>3.590000e-56</td>\n",
       "      <td>191.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH14-bin.2</th>\n",
       "      <td>gene1003</td>\n",
       "      <td>9.490000e-23</td>\n",
       "      <td>99.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH9-bin.1</th>\n",
       "      <td>gene3888</td>\n",
       "      <td>1.190000e-22</td>\n",
       "      <td>99.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.10</th>\n",
       "      <td>gene2510</td>\n",
       "      <td>4.490000e-19</td>\n",
       "      <td>87.0</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH9-bin.8</th>\n",
       "      <td>gene1</td>\n",
       "      <td>1.140000e-18</td>\n",
       "      <td>84.3</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.18</th>\n",
       "      <td>gene948</td>\n",
       "      <td>2.010000e-17</td>\n",
       "      <td>82.4</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.14</th>\n",
       "      <td>gene5876</td>\n",
       "      <td>2.620000e-17</td>\n",
       "      <td>82.4</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.11</th>\n",
       "      <td>gene2382</td>\n",
       "      <td>4.360000e-17</td>\n",
       "      <td>80.5</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.2</th>\n",
       "      <td>gene979</td>\n",
       "      <td>4.580000e-17</td>\n",
       "      <td>80.5</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.12</th>\n",
       "      <td>gene529</td>\n",
       "      <td>5.030000e-17</td>\n",
       "      <td>80.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.11</th>\n",
       "      <td>gene2414</td>\n",
       "      <td>8.090000e-17</td>\n",
       "      <td>79.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.4</th>\n",
       "      <td>gene68</td>\n",
       "      <td>8.480000e-17</td>\n",
       "      <td>78.6</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.14</th>\n",
       "      <td>gene546</td>\n",
       "      <td>9.540000e-17</td>\n",
       "      <td>79.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.0</th>\n",
       "      <td>gene2567</td>\n",
       "      <td>1.030000e-16</td>\n",
       "      <td>79.3</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.24</th>\n",
       "      <td>gene385</td>\n",
       "      <td>6.930000e-16</td>\n",
       "      <td>75.5</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.13</th>\n",
       "      <td>gene2195</td>\n",
       "      <td>1.170000e-15</td>\n",
       "      <td>75.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.4</th>\n",
       "      <td>gene2613</td>\n",
       "      <td>1.580000e-15</td>\n",
       "      <td>75.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.1</th>\n",
       "      <td>gene2599</td>\n",
       "      <td>1.620000e-15</td>\n",
       "      <td>75.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.16</th>\n",
       "      <td>gene1114</td>\n",
       "      <td>2.510000e-15</td>\n",
       "      <td>75.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.13</th>\n",
       "      <td>gene458</td>\n",
       "      <td>2.780000e-15</td>\n",
       "      <td>74.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.9</th>\n",
       "      <td>gene614</td>\n",
       "      <td>3.530000e-15</td>\n",
       "      <td>73.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH3-bin.1</th>\n",
       "      <td>gene4869</td>\n",
       "      <td>3.850000e-15</td>\n",
       "      <td>75.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.23</th>\n",
       "      <td>gene1853</td>\n",
       "      <td>7.300000e-15</td>\n",
       "      <td>73.6</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.27</th>\n",
       "      <td>gene588</td>\n",
       "      <td>7.450000e-15</td>\n",
       "      <td>73.6</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.20</th>\n",
       "      <td>gene898</td>\n",
       "      <td>2.310000e-14</td>\n",
       "      <td>70.9</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH3-bin.2</th>\n",
       "      <td>gene2757</td>\n",
       "      <td>2.650000e-14</td>\n",
       "      <td>72.4</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH14-bin.1</th>\n",
       "      <td>gene1344</td>\n",
       "      <td>6.140000e-14</td>\n",
       "      <td>71.2</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.2</th>\n",
       "      <td>gene1742</td>\n",
       "      <td>6.680000e-14</td>\n",
       "      <td>70.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.8</th>\n",
       "      <td>gene2703</td>\n",
       "      <td>1.350000e-13</td>\n",
       "      <td>69.7</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.15</th>\n",
       "      <td>gene1972</td>\n",
       "      <td>3.030000e-13</td>\n",
       "      <td>68.6</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH1-bin.0</th>\n",
       "      <td>gene2609</td>\n",
       "      <td>3.480000e-13</td>\n",
       "      <td>68.2</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.7</th>\n",
       "      <td>gene621</td>\n",
       "      <td>3.000000e-12</td>\n",
       "      <td>65.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.25</th>\n",
       "      <td>gene855</td>\n",
       "      <td>3.060000e-12</td>\n",
       "      <td>65.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH1-bin.4</th>\n",
       "      <td>gene2269</td>\n",
       "      <td>8.890000e-12</td>\n",
       "      <td>63.2</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.20</th>\n",
       "      <td>gene1181</td>\n",
       "      <td>4.790000e-11</td>\n",
       "      <td>61.2</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.10</th>\n",
       "      <td>gene75</td>\n",
       "      <td>7.650000e-11</td>\n",
       "      <td>60.8</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.21</th>\n",
       "      <td>gene1580</td>\n",
       "      <td>1.220000e-10</td>\n",
       "      <td>60.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.0</th>\n",
       "      <td>gene1211</td>\n",
       "      <td>1.460000e-10</td>\n",
       "      <td>58.5</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.16</th>\n",
       "      <td>gene888</td>\n",
       "      <td>1.740000e-10</td>\n",
       "      <td>58.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.9</th>\n",
       "      <td>gene1470</td>\n",
       "      <td>4.890000e-10</td>\n",
       "      <td>58.5</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH1-bin.8</th>\n",
       "      <td>gene2827</td>\n",
       "      <td>1.340000e-08</td>\n",
       "      <td>54.3</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.24</th>\n",
       "      <td>gene534</td>\n",
       "      <td>2.740000e-08</td>\n",
       "      <td>51.6</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH7-bin.25</th>\n",
       "      <td>gene808</td>\n",
       "      <td>3.460000e-07</td>\n",
       "      <td>48.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.17</th>\n",
       "      <td>gene51</td>\n",
       "      <td>3.780000e-07</td>\n",
       "      <td>48.9</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.0</th>\n",
       "      <td>gene1303</td>\n",
       "      <td>6.750000e-07</td>\n",
       "      <td>48.1</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH13-bin.17</th>\n",
       "      <td>gene572</td>\n",
       "      <td>8.200000e-07</td>\n",
       "      <td>47.8</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH8-bin.5</th>\n",
       "      <td>gene1101</td>\n",
       "      <td>8.840000e-07</td>\n",
       "      <td>47.8</td>\n",
       "      <td>2.A.21.7.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH15-bin.13</th>\n",
       "      <td>gene1023</td>\n",
       "      <td>4.300000e-06</td>\n",
       "      <td>45.4</td>\n",
       "      <td>2.A.1.13.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene         evalue  bit_score     TCDB_id\n",
       "MAG                                                        \n",
       "CH15-bin.17   gene545  4.900000e-157      456.0  2.A.21.7.3\n",
       "CH13-bin.1    gene374  4.910000e-157      456.0  2.A.21.7.3\n",
       "CH7-bin.16   gene2185  8.200000e-156      452.0  2.A.21.7.3\n",
       "CH7-bin.4    gene3341  4.480000e-153      445.0  2.A.21.7.3\n",
       "CH8-bin.25   gene3105  2.220000e-151      441.0  2.A.21.7.3\n",
       "CH1-bin.9     gene769  1.280000e-150      439.0  2.A.21.7.3\n",
       "CH15-bin.6   gene1704  1.430000e-142      420.0  2.A.21.7.3\n",
       "CH1-bin.1     gene727   3.590000e-56      191.0  2.A.21.7.3\n",
       "CH14-bin.2   gene1003   9.490000e-23       99.0  2.A.21.7.3\n",
       "CH9-bin.1    gene3888   1.190000e-22       99.0  2.A.21.7.3\n",
       "CH13-bin.10  gene2510   4.490000e-19       87.0  2.A.21.7.3\n",
       "CH9-bin.8       gene1   1.140000e-18       84.3  2.A.1.13.1\n",
       "CH15-bin.18   gene948   2.010000e-17       82.4  2.A.21.7.3\n",
       "CH13-bin.14  gene5876   2.620000e-17       82.4  2.A.21.7.3\n",
       "CH7-bin.11   gene2382   4.360000e-17       80.5  2.A.21.7.3\n",
       "CH8-bin.2     gene979   4.580000e-17       80.5  2.A.21.7.3\n",
       "CH8-bin.12    gene529   5.030000e-17       80.1  2.A.21.7.3\n",
       "CH8-bin.11   gene2414   8.090000e-17       79.7  2.A.21.7.3\n",
       "CH8-bin.4      gene68   8.480000e-17       78.6  2.A.1.13.1\n",
       "CH8-bin.14    gene546   9.540000e-17       79.7  2.A.21.7.3\n",
       "CH13-bin.0   gene2567   1.030000e-16       79.3  2.A.21.7.3\n",
       "CH7-bin.24    gene385   6.930000e-16       75.5  2.A.1.13.1\n",
       "CH7-bin.13   gene2195   1.170000e-15       75.9  2.A.21.7.3\n",
       "CH13-bin.4   gene2613   1.580000e-15       75.9  2.A.21.7.3\n",
       "CH15-bin.1   gene2599   1.620000e-15       75.9  2.A.21.7.3\n",
       "CH8-bin.16   gene1114   2.510000e-15       75.1  2.A.21.7.3\n",
       "CH13-bin.13   gene458   2.780000e-15       74.7  2.A.21.7.3\n",
       "CH8-bin.9     gene614   3.530000e-15       73.9  2.A.21.7.3\n",
       "CH3-bin.1    gene4869   3.850000e-15       75.1  2.A.21.7.3\n",
       "CH15-bin.23  gene1853   7.300000e-15       73.6  2.A.21.7.3\n",
       "CH8-bin.27    gene588   7.450000e-15       73.6  2.A.21.7.3\n",
       "CH7-bin.20    gene898   2.310000e-14       70.9  2.A.1.13.1\n",
       "CH3-bin.2    gene2757   2.650000e-14       72.4  2.A.21.7.3\n",
       "CH14-bin.1   gene1344   6.140000e-14       71.2  2.A.21.7.3\n",
       "CH13-bin.2   gene1742   6.680000e-14       70.9  2.A.21.7.3\n",
       "CH7-bin.8    gene2703   1.350000e-13       69.7  2.A.21.7.3\n",
       "CH7-bin.15   gene1972   3.030000e-13       68.6  2.A.21.7.3\n",
       "CH1-bin.0    gene2609   3.480000e-13       68.2  2.A.21.7.3\n",
       "CH15-bin.7    gene621   3.000000e-12       65.1  2.A.21.7.3\n",
       "CH13-bin.25   gene855   3.060000e-12       65.1  2.A.21.7.3\n",
       "CH1-bin.4    gene2269   8.890000e-12       63.2  2.A.1.13.1\n",
       "CH13-bin.20  gene1181   4.790000e-11       61.2  2.A.21.7.3\n",
       "CH15-bin.10    gene75   7.650000e-11       60.8  2.A.21.7.3\n",
       "CH7-bin.21   gene1580   1.220000e-10       60.1  2.A.21.7.3\n",
       "CH8-bin.0    gene1211   1.460000e-10       58.5  2.A.1.13.1\n",
       "CH13-bin.16   gene888   1.740000e-10       58.9  2.A.21.7.3\n",
       "CH7-bin.9    gene1470   4.890000e-10       58.5  2.A.21.7.3\n",
       "CH1-bin.8    gene2827   1.340000e-08       54.3  2.A.21.7.3\n",
       "CH13-bin.24   gene534   2.740000e-08       51.6  2.A.1.13.1\n",
       "CH7-bin.25    gene808   3.460000e-07       48.9  2.A.21.7.3\n",
       "CH8-bin.17     gene51   3.780000e-07       48.9  2.A.21.7.3\n",
       "CH15-bin.0   gene1303   6.750000e-07       48.1  2.A.21.7.3\n",
       "CH13-bin.17   gene572   8.200000e-07       47.8  2.A.21.7.3\n",
       "CH8-bin.5    gene1101   8.840000e-07       47.8  2.A.21.7.3\n",
       "CH15-bin.13  gene1023   4.300000e-06       45.4  2.A.1.13.1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ACt2r_MAGs_df = pd.DataFrame(ACt2r_MAGs_data,columns=[\"MAG\",\"gene\",\"evalue\",\"bit_score\",\"TCDB_id\"]).sort_values(\"evalue\")\n",
    "ACt2r_MAGs_df.set_index(\"MAG\",inplace=True)\n",
    "ACt2r_MAGs_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba715e30-0a58-4373-81b8-585945d37fcd",
   "metadata": {},
   "source": [
    "### Add acetate transport\n",
    "\n",
    "Two different conditions.\n",
    "\n",
    "1. The MAG has a hit in the TCDB database -> add reversible reaction\n",
    "2. The MAG has a hit for the enzymes in acetate production -> add only producing reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3e7b7ec4-f53e-4d44-bf81-0884ce91ffb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "GEMs_dict2 = {}\n",
    "\n",
    "for MAG,model in GEMs_dict.items():\n",
    "    model_copy = model.copy()\n",
    "    \n",
    "    if \"R_Acabc\" in model_copy.reactions.keys(): \n",
    "        model_copy.remove_reaction(\"R_Acabc\")\n",
    "      \n",
    "    if \"R_ACt2r\" in model_copy.reactions.keys():\n",
    "        model_copy.remove_reaction(\"R_ACt2r\")\n",
    "        \n",
    "    if MAG in ACt2r_MAGs_df.index.values:\n",
    "        model_copy.add_reaction_from_str(\"R_ACt2r: M_ac_c + M_h_c --> M_ac_e + M_h_e\")\n",
    "        model_copy.reversible=True\n",
    "        model_copy.lb=-1000\n",
    "        \n",
    "        GPR = parse_gpr_rule(ACt2r_MAGs_df.loc[MAG,\"gene\"])\n",
    "        model_copy.set_gpr_association(\"R_ACt2r\",GPR)\n",
    "        \n",
    "        \n",
    "    elif has_enzymes_for_acetate_MAGs:\n",
    "        model_copy.add_reaction_from_str(\"R_ACt2r: M_ac_c + M_h_c --> M_ac_e + M_h_e\")\n",
    "        \n",
    "    GEMs_dict2[MAG]=model_copy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d0f41c7-81e8-48d8-889a-1fe610d1d064",
   "metadata": {},
   "source": [
    "### Look for energy generating cycles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "52bfe792-ccb0-41c0-ad51-20051bf1902f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import functions.EGC as EGC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0b663cc6-7a5c-4fb0-8e27-964504ab087e",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CH7-bin.18\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-03-14\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.7\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.0\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH1-bin.4\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.1\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.11\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.6\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.2\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.13\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH1-bin.10\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH1-bin.6\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.2\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.12\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.5\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.1\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.1\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH1-bin.3\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.17\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.0\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.4\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.23\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.6\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.22\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.2\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH1-bin.1\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.4\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.14\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.23\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.20\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.8\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.9\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.25\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH14-bin.4\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH14-bin.1\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.21\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH14-bin.2\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.22\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.2\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.29\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.7\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.6\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH3-bin.2\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.14\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.16\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH3-bin.0\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.5\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH3-bin.1\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH8-bin.17\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.11\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH9-bin.1\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.12\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH13-bin.25\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.13\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH9-bin.0\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.12\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH9-bin.2\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.9\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.10\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.8\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.13\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.17\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.8\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.15\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.16\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH9-bin.6\n",
      "\u001b[91mThere ARE energy producing cycles in the model\u001b[0m\n",
      "CH9-bin.4\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.17\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH1-bin.8\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH1-bin.9\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH15-bin.16\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH7-bin.15\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n",
      "CH9-bin.5\n",
      "\u001b[91mThere ARE energy producing cycles in the model\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "for MAG,model in GEMs_dict2.items():\n",
    "    print(MAG)\n",
    "    EGC.EGC_identifier(model,print_results=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "353c6b25-d312-4b5d-86bf-adb544a29063",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_uni = reframed.load_cbmodel(\"/Users/idunmariaburgos/universal_model_extension/output/universe_bacteria.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "645f9f76-88da-45cb-8adf-efc8074ec6c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are NO energy producing cycles in the model for R_ATPM\n",
      "There are NO energy producing cycles in the model for R_CTPM\n",
      "There are NO energy producing cycles in the model for R_GTPM\n",
      "There are NO energy producing cycles in the model for R_UTPM\n",
      "There are NO energy producing cycles in the model for R_ITPM\n",
      "There are NO energy producing cycles in the model for R_nadhM\n",
      "There are NO energy producing cycles in the model for R_nadphM\n",
      "There are NO energy producing cycles in the model for R_fadh2M\n",
      "There are NO energy producing cycles in the model for R_fmnh2M\n",
      "There are NO energy producing cycles in the model for R_q8h2M\n",
      "There are NO energy producing cycles in the model for R_mql8M\n",
      "There are NO energy producing cycles in the model for R_dmmql8M\n",
      "There are NO energy producing cycles in the model for R_accoaM\n",
      "There are NO energy producing cycles in the model for R_gluM\n",
      "There are NO energy producing cycles in the model for R_protonM\n",
      "There are NO energy producing cycles in the model for R_ferredoxinM\n",
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGC.EGC_identifier(model_uni,print_results=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "268367d7-bb75-4f85-945f-be9c3a6ef4bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are energy producing cycles in the model for R_ATPM\n",
      "\tR_ALAD_L: 1000.0\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_CYTBO3_4pp: 1000.0\n",
      "\tR_GLYCLTDx: 1000.0\n",
      "\tR_GLYCTO2: 1000.0\n",
      "\tR_ATPM: 1000.0\n",
      "\tR_CAT: 1000.0\n",
      "\tR_PDYXPT_c: 1000.0\n",
      "\tR_PYDXO: 1000.0\n",
      "There are energy producing cycles in the model for R_CTPM\n",
      "\tR_ALAD_L: 1000.0\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_CYTBO3_4pp: 1000.0\n",
      "\tR_CYTK2: 1000.0\n",
      "\tR_CYTK2_1: -1000.0\n",
      "\tR_GLYCLTDx: 1000.0\n",
      "\tR_GLYCTO2: 1000.0\n",
      "\tR_CAT: 1000.0\n",
      "\tR_PDYXPT_c: 1000.0\n",
      "\tR_PYDXO: 1000.0\n",
      "\tR_CTPM: 1000.0\n",
      "There are energy producing cycles in the model for R_GTPM\n",
      "\tR_ADOCBIK: 1000.0\n",
      "\tR_ALAD_L: 1000.0\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_CYTBO3_4pp: 1000.0\n",
      "\tR_GDOCBIK: -1000.0\n",
      "\tR_GLYCLTDx: 1000.0\n",
      "\tR_GLYCTO2: 1000.0\n",
      "\tR_CAT: 1000.0\n",
      "\tR_PDYXPT_c: 1000.0\n",
      "\tR_PYDXO: 1000.0\n",
      "\tR_GTPM: 1000.0\n",
      "There are energy producing cycles in the model for R_UTPM\n",
      "\tR_ALAD_L: 1000.0\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_CYTBO3_4pp: 1000.0\n",
      "\tR_GLYCLTDx: 1000.0\n",
      "\tR_GLYCTO2: 1000.0\n",
      "\tR_NDPK2: 1000.0\n",
      "\tR_CAT: 1000.0\n",
      "\tR_PDYXPT_c: 1000.0\n",
      "\tR_PYDXO: 1000.0\n",
      "\tR_UTPM: 1000.0\n",
      "There are energy producing cycles in the model for R_ITPM\n",
      "\tR_ALAD_L: 1000.0000000000002\n",
      "\tR_ATPS4rpp: 1000.0000000000002\n",
      "\tR_CYTBO3_4pp: 1000.0000000000002\n",
      "\tR_GLYCLTDx: 1000.0000000000002\n",
      "\tR_GLYCTO2: 1000.0000000000002\n",
      "\tR_NDPK9: 1000.0\n",
      "\tR_CAT: 1000.0000000000002\n",
      "\tR_PDYXPT_c: 1000.0000000000002\n",
      "\tR_PYDXO: 1000.0000000000002\n",
      "\tR_ITPM: 1000.0\n",
      "There are NO energy producing cycles in the model for R_nadhM\n",
      "There are NO energy producing cycles in the model for R_nadphM\n",
      "There are NO energy producing cycles in the model for R_fadh2M\n",
      "There are NO energy producing cycles in the model for R_fmnh2M\n",
      "There are NO energy producing cycles in the model for R_q8h2M\n",
      "There are NO energy producing cycles in the model for R_mql8M\n",
      "There are NO energy producing cycles in the model for R_dmmql8M\n",
      "There are energy producing cycles in the model for R_accoaM\n",
      "\tR_ACKr: 1000.0\n",
      "\tR_ALAD_L: 1000.0\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_CYTBO3_4pp: 1000.0\n",
      "\tR_GLYCLTDx: 1000.0\n",
      "\tR_GLYCTO2: 1000.0\n",
      "\tR_PTAr: -1000.0\n",
      "\tR_CAT: 1000.0\n",
      "\tR_PDYXPT_c: 1000.0\n",
      "\tR_PYDXO: 1000.0\n",
      "\tR_accoaM: 1000.0\n",
      "There are NO energy producing cycles in the model for R_gluM\n",
      "There are energy producing cycles in the model for R_protonM\n",
      "\tR_ALAD_L: 250.0\n",
      "\tR_CYTBO3_4pp: 250.0\n",
      "\tR_GLYCLTDx: 250.0\n",
      "\tR_GLYCTO2: 250.0\n",
      "\tR_CAT: 250.0\n",
      "\tR_PDYXPT_c: 250.0\n",
      "\tR_PYDXO: 250.0\n",
      "\tR_protonM: 1000.0\n",
      "There are NO energy producing cycles in the model for R_ferredoxinM\n",
      "\u001b[91mThere ARE energy producing cycles in the model\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'R_ALAD_L',\n",
       " 'R_CAT',\n",
       " 'R_CYTBO3_4pp',\n",
       " 'R_GLYCLTDx',\n",
       " 'R_GLYCTO2',\n",
       " 'R_PDYXPT_c',\n",
       " 'R_PYDXO'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGC_results = EGC.EGC_identifier(GEMs_dict2[\"CH9-bin.5\"],print_results=True)\n",
    "set.intersection(*map(set,[[value[0] for value in values] for source,values in EGC_results.items()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f20f4ae7-67d6-44f0-a0d4-b77dc0707c69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R_PYDXO: 2.0 M_h2o_c + M_nh4_c + 0.5 M_o2_c + M_pydx_c <-> 2.0 M_h2o2_c + M_pydam_c [-inf, 0.0]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_uni.reactions.R_PYDXO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5bf0d029-2a2a-4717-824e-2039f4599dd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R_PYDXO: 2.0 M_h2o_c + M_nh4_c + 0.5 M_o2_c + M_pydx_c <-> 2.0 M_h2o2_c + M_pydam_c"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GEMs_dict2[\"CH9-bin.5\"].reactions.R_PYDXO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d7b4cbf7-6c31-4e95-bbfb-8ce323d80b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_9_5 = reframed.load_cbmodel(\"output/GEMs/GEMs_adapt/CH9-bin.5.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6027d13c-6514-4a0e-8fd6-951a5fd7016a",
   "metadata": {},
   "outputs": [],
   "source": [
    "GEMs_dict2[\"CH9-bin.5\"].remove_reactions(list(set(GEMs_dict2[\"CH9-bin.5\"].reactions)-set(model_9_5.reactions)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9219451e-d721-449a-908f-1c350307efae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "EGC_results = EGC.EGC_identifier(GEMs_dict2[\"CH9-bin.5\"],print_results=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "afde96a6-9817-419b-8d58-68e1c12cf154",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are energy producing cycles in the model for R_ATPM\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_MG2tex: 2000.0\n",
      "\tR_MGt5: 2000.0\n",
      "\tR_ATPM: 1000.0\n",
      "\tR_MG2t3_2pp: 2000.0\n",
      "There are energy producing cycles in the model for R_CTPM\n",
      "\tR_ADK1: 1000.0\n",
      "\tR_ATPS4rpp: 2000.0\n",
      "\tR_MG2tex: 4000.0\n",
      "\tR_MGt5: 4000.0\n",
      "\tR_PPA: 1000.0\n",
      "\tR_PPDK: 1000.0\n",
      "\tR_PYK4: 1000.0\n",
      "\tR_MG2t3_2pp: 4000.0\n",
      "\tR_CTPM: 1000.0\n",
      "There are energy producing cycles in the model for R_GTPM\n",
      "\tR_ADK1: 1000.0\n",
      "\tR_ADK3: -1000.0\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_MG2tex: 2000.0\n",
      "\tR_MGt5: 2000.0\n",
      "\tR_MG2t3_2pp: 2000.0\n",
      "\tR_GTPM: 1000.0\n",
      "There are energy producing cycles in the model for R_UTPM\n",
      "\tR_ADK1: 1000.0\n",
      "\tR_ATPS4rpp: 2000.0\n",
      "\tR_MG2tex: 4000.0\n",
      "\tR_MGt5: 4000.0\n",
      "\tR_PPA: 1000.0\n",
      "\tR_PPDK: 1000.0\n",
      "\tR_PYK2: 1000.0\n",
      "\tR_MG2t3_2pp: 4000.0\n",
      "\tR_UTPM: 1000.0\n",
      "There are energy producing cycles in the model for R_ITPM\n",
      "\tR_ADK1: 1000.0\n",
      "\tR_ADK4: -1000.0\n",
      "\tR_ATPS4rpp: 1000.0\n",
      "\tR_MG2tex: 2000.0\n",
      "\tR_MGt5: 2000.0\n",
      "\tR_MG2t3_2pp: 2000.0\n",
      "\tR_ITPM: 1000.0\n",
      "There are NO energy producing cycles in the model for R_nadhM\n",
      "There are NO energy producing cycles in the model for R_nadphM\n",
      "There are NO energy producing cycles in the model for R_fadh2M\n",
      "There are NO energy producing cycles in the model for R_fmnh2M\n",
      "There are NO energy producing cycles in the model for R_q8h2M\n",
      "There are NO energy producing cycles in the model for R_mql8M\n",
      "There are NO energy producing cycles in the model for R_accoaM\n",
      "There are NO energy producing cycles in the model for R_gluM\n",
      "There are energy producing cycles in the model for R_protonM\n",
      "\tR_MG2tex: 500.0\n",
      "\tR_MGt5: 500.0\n",
      "\tR_MG2t3_2pp: 500.0\n",
      "\tR_protonM: 1000.0\n",
      "There are NO energy producing cycles in the model for R_ferredoxinM\n",
      "\u001b[91mThere ARE energy producing cycles in the model\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'R_MG2t3_2pp', 'R_MG2tex', 'R_MGt5'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGC_results = EGC.EGC_identifier(GEMs_dict2[\"CH9-bin.6\"],print_results=True)\n",
    "set.intersection(*map(set,[[value[0] for value in values] for source,values in EGC_results.items()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "30a357bc-48ff-4526-864b-6f8e9f282eaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R_MG2t3_2pp: 2.0 M_h_c + M_mg2_p <-> 2.0 M_h_p + M_mg2_c [-inf, 0.0]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_uni.reactions.R_MG2t3_2pp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "05a427e4-0c5a-4a4a-8629-162894863e99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R_MG2t3_2pp: 2.0 M_h_c + M_mg2_p <-> 2.0 M_h_p + M_mg2_c"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GEMs_dict2[\"CH9-bin.6\"].reactions.R_MG2t3_2pp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8ba2fa3-d414-476a-ba40-706fbe3c4c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_9_6 = reframed.load_cbmodel(\"output/GEMs/GEMs_adapt/CH9-bin.6.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "eaa79859-837c-4ed0-9814-702f5d4632f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "GEMs_dict2[\"CH9-bin.6\"].remove_reactions(list(set(GEMs_dict2[\"CH9-bin.6\"].reactions)-set(model_9_6.reactions)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7e56a8b9-9762-4bbd-8e93-27e07fd83af5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92mThere are NO energy producing cycles in the model\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EGC.EGC_identifier(GEMs_dict2[\"CH9-bin.6\"],print_results=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "aa6cf1d3-6825-482f-84e1-04dd25a419c8",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/idunmariaburgos/opt/anaconda3/envs/MAGs_RecAndAn2/lib/python3.9/site-packages/reframed/core/model.py:471: UserWarning: No such reaction R_EX_o2_e\n",
      "  warn(f\"No such reaction {r_id}\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n",
      "['R_O2tex', 'R_EX_o2_e']\n"
     ]
    }
   ],
   "source": [
    "GEMs_dict3 = {}\n",
    "for MAG, model in GEMs_dict2.items():\n",
    "    \n",
    "    model_copy = model.copy()\n",
    "    \n",
    "    print(model_copy.get_metabolite_reactions(\"M_o2_e\"))\n",
    "    model_copy.remove_reactions(model_copy.get_metabolite_reactions(\"M_o2_e\"))\n",
    "    model_copy.remove_metabolite(\"M_o2_e\")\n",
    "    model_copy.remove_reaction(\"R_EX_o2_e\")\n",
    "    \n",
    "    GEMs_dict3[MAG]=model_copy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afb2b98b-fb2f-4d00-aef5-099f169373f3",
   "metadata": {},
   "source": [
    "### Save models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cc837bfd-2eb4-4e2b-a93c-8cd6e9a2fec3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for MAG,model in GEMs_dict3.items():\n",
    "    model.update()\n",
    "    reframed.save_cbmodel(model,\"output/GEMs/GEMs_adapt_media_ACt2r/\"+MAG+\".xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d3a354d-f9d4-423e-86d3-5bdde9761fc2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MAGs_RecAndAn2",
   "language": "python",
   "name": "mags_recandan2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
